[jit][script] enable advanced indexing with tensors

zou3519 opened the pull request:
On the way to #10774

This PR adds advanced indexing with tensors.
The approach is to desugar advanced indexing into an at::index op.
This is exactly how normal pytorch does it.
[(I used this code as reference)](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_variable_indexing.cpp)

Supporting sequences is a little tricky because JIT script doesn't have
an easy way to turn arbitrary n-dimensional python lists into a tensor
(it would be easy if we supported `torch.tensor`), so that'll come
in a future PR.

cc @jamesr66a @zdevito

zou3519 committed:
[jit] enable advanced indexing with tensors

On the way to #10774

This PR adds advanced indexing with tensors.
The approach is to desugar advanced indexing into an at::index op.
This is exactly how normal pytorch does it.
[(I used this code as reference)](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_variable_indexing.cpp)

Sequences are a little tricky because JIT script doesn't support
an easy way to turn arbitrary n-dimensional python lists into a tensor
(it would be easy if we supported `torch.tensor`), so that'll come
in a future PR.

apaszke commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1616,6 +1625,16 @@ struct to_ir {
+    auto handle_tensor = [&](Value* tensor) {
+      tensor_indices.resize(dim + 1);
+      tensor_indices[dim] = tensor;
+      dim++;
```
How is that different than `tensor_indices.push_back(tensor); dim++`? If the vector can have holes can you add a comment saying how exactly those indices should be understood? Do they correspond to indexing before or after processing the slices?

apaszke replied:
Also, it seems to be used in a single place only, so it shouldn't be necessary.

zou3519 replied:
Yes, the vector can have holes in it. I'll add a comment saying how those are understood (`at::index` requires holes at indexes that aren't being indexed). 

handle_tensor is used in one place but in the future it'll be used in more places. Supporting indexing with sequences can be implemented by converting the sequence to a tensor and then using `at::index` on it.

apaszke commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1668,11 +1704,18 @@ struct to_ir {
+            loc, gatherable.value(*graph), /*dim=*/0, index);
+      }
+      throw ErrorReport(loc)
+        << "Cannot index tensor with value of type " << index_type->str();
```
It's really annoying that we have this thing implemented in multiple places. Can't we just call `emitMultidimIndexing` in here?

zou3519 replied:
Can do. The code is a little annoying to work around because emitMultidimIndex assumes that only the `gatherable` has been emitted, but `emitIndex` emits both the `gatherable` and the `index` and I would have to tell `emitIndex` to only emit the `gatherable`. I'll do some refactoring.

zou3519 committed:
Address comments:
- Add note about what the holes in tensor_indices are
- Do some refactoring to make the code cleaner

zdevito reviewed (approved):
This looks good to me.

zou3519 committed:
fix rebase
